File: iceoryx2-ffi/python/tests/node_tests.py
# Copyright (c) 2025 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Apache Software License 2.0 which is available at
# https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
# which is available at https://opensource.org/licenses/MIT.
#
# SPDX-License-Identifier: Apache-2.0 OR MIT

import iceoryx2 as iox2
import pytest

service_types = [iox2.ServiceType.Ipc, iox2.ServiceType.Local]


@pytest.mark.parametrize("service_type", service_types)
def test_creating_node_works(service_type: iox2.ServiceType) -> None:
    sut = iox2.NodeBuilder.new().create(service_type)
    assert sut.name.as_str() == ""


@pytest.mark.parametrize("service_type", service_types)
def test_creating_node_with_properties_works(
    service_type: iox2.ServiceType,
) -> None:
    node_name = iox2.testing.generate_node_name()
    signal_handling_mode = iox2.SignalHandlingMode.Disabled
    config = iox2.testing.generate_isolated_config()

    sut = (
        iox2.NodeBuilder.new()
        .name(node_name)
        .signal_handling_mode(signal_handling_mode)
        .config(config)
        .create(service_type)
    )
    assert sut.name == node_name
    assert sut.signal_handling_mode == signal_handling_mode
    assert sut.config == config


@pytest.mark.parametrize("service_type", service_types)
def test_cleanup_dead_nodes_can_be_called(
    service_type: iox2.ServiceType,
) -> None:
    try:
        iox2.Node.cleanup_dead_nodes(
            service_type, iox2.testing.generate_isolated_config()
        )
    except iox2.NodeCleanupFailure:
        assert False


@pytest.mark.parametrize("service_type", service_types)
def test_created_nodes_can_be_listed(service_type: iox2.ServiceType) -> None:
    config = iox2.testing.generate_isolated_config()
    sut_1 = (
        iox2.NodeBuilder.new()
        .name(iox2.NodeName.new("behind you, there is"))
        .config(config)
        .create(service_type)
    )
    sut_2 = (
        iox2.NodeBuilder.new()
        .name(iox2.NodeName.new("a 3 headed monkey"))
        .config(config)
        .create(service_type)
    )

    node_list = iox2.Node.list(service_type, config)
    assert len(node_list) == 2
    for node in node_list:
        match node:
            case node.Alive():
                assert isinstance(node, iox2.NodeState.Alive)
                assert node[0].id in (sut_1.id, sut_2.id)
                assert node[0].details.name in (sut_1.name, sut_2.name)
            case node.Dead():
                assert False
            case node.Inaccessible():
                assert False
            case node.Undefined():
                assert False


@pytest.mark.parametrize("service_type", service_types)
def test_wait_can_be_called(service_type: iox2.ServiceType) -> None:
    config = iox2.testing.generate_isolated_config()
    sut = iox2.NodeBuilder.new().config(config).create(service_type)
    try:
        sut.wait(iox2.Duration.from_millis(1))
    except iox2.NodeWaitFailure:
        assert False

File: iceoryx2-cxx/tests/src/node_tests.cpp
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include "iox2/node.hpp"
#include "iox2/node_name.hpp"

#include <vector>

#include "test.hpp"

namespace {
using namespace iox2;

template <typename T>
class NodeTest : public ::testing::Test {
  public:
    static constexpr ServiceType TYPE = T::TYPE;
};

TYPED_TEST_SUITE(NodeTest, iox2_testing::ServiceTypes, );

TYPED_TEST(NodeTest, node_name_is_applied) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    const auto* name_value = "First time we met, I saw the ocean, it was wet!";
    auto node_name = NodeName::create(name_value).expect("");

    auto sut = NodeBuilder().name(node_name).create<SERVICE_TYPE>().expect("");
    ASSERT_THAT(sut.name().to_string(), Eq(node_name.to_string()));
}

TYPED_TEST(NodeTest, created_nodes_can_be_listed) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    auto node_name_1 = NodeName::create("Nala does not like water.").expect("");
    auto node_name_2 = NodeName::create("Nala does not like paprika.").expect("");

    {
        auto sut_1 = NodeBuilder().name(node_name_1).create<SERVICE_TYPE>().expect("");
        auto sut_2 = NodeBuilder().name(node_name_2).create<SERVICE_TYPE>().expect("");

        std::vector<NodeName> nodes;
        auto result = Node<SERVICE_TYPE>::list(Config::global_config(), [&](auto node_state) {
            node_state.alive([&](auto& view) { nodes.push_back(view.details()->name()); });
            return CallbackProgression::Continue;
        });
        ASSERT_TRUE(result.has_value());

        auto contains = [&](const NodeName& name) {
            // NOLINTNEXTLINE(readability-use-anyofallof), not yet supported in all compilers
            for (const auto& node : nodes) {
                if (node.to_string() == name.to_string()) {
                    return true;
                }
            }
            return false;
        };

        ASSERT_TRUE(contains(node_name_1));
        ASSERT_TRUE(contains(node_name_2));
    }

    uint64_t counter = 0;
    auto result = Node<SERVICE_TYPE>::list(Config::global_config(), [&](const auto&) {
        counter++;
        return CallbackProgression::Continue;
    });
    ASSERT_TRUE(result.has_value());
    ASSERT_THAT(counter, Eq(0));
}

TYPED_TEST(NodeTest, signal_handling_mode_can_be_set) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    auto sut_1 = NodeBuilder().signal_handling_mode(SignalHandlingMode::Disabled).create<SERVICE_TYPE>().expect("");
    auto sut_2 = NodeBuilder()
                     .signal_handling_mode(SignalHandlingMode::HandleTerminationRequests)
                     .create<SERVICE_TYPE>()
                     .expect("");

    ASSERT_THAT(sut_1.signal_handling_mode(), Eq(SignalHandlingMode::Disabled));
    ASSERT_THAT(sut_2.signal_handling_mode(), Eq(SignalHandlingMode::HandleTerminationRequests));
}

TYPED_TEST(NodeTest, node_id_is_unique) {
    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;

    auto sut_1 = NodeBuilder().create<SERVICE_TYPE>().expect("");
    auto sut_2 = NodeBuilder().create<SERVICE_TYPE>().expect("");

    auto id_1 = sut_1.id();
    auto id_1_1 = sut_1.id();
    auto id_2 = sut_2.id();

    ASSERT_THAT(id_1, Eq(id_1_1));
    ASSERT_THAT(id_2, Ne(id_1));
    ASSERT_THAT(id_1.pid(), Eq(id_2.pid()));
}
} // namespace
